docs/developer-docs/backend/rust/1-infrastructure.md

```diff
@@ -0,0 +1,204 @@
+
+
+### Using `static mut STATE: Option<State> = None;`
+
+This approach uses plain global variables and forces you to write some boilerplate to access the global state. This method should also be used with caution, since it also allows you to obtain multiple non-exclusive mutable references. 
+
+### Canister code should be target-independent 
+
+It pays off to factor most of the canister code into loosely coupled modules and packages and to test them independently. Most of the code that depends on the System API should go into the main file.
+
+It is also possible to create a thin abstractions for the System API and test your code with a fake but faithful implementation. For example, we could use the following trait to abstract the stable memory API:
+
+```
+pub trait Memory {
+    fn size(&self) -> WasmPages;
+    fn grow(&self, pages: WasmPages) -> WasmPages;
+    fn read(&self, offset: u32, dst: &mut [u8]);
+    fn write(&self, offset: u32, src: &[u8]);
+}
+```
+
+## Code asynchrony
+
+### Traps and panics
+In canisters, panics and traps are somewhat unique. If a code panics or traps, the system rolls the canister's state back to the latest working snapshot. This means that if your canister makes a call and then panicked in the callback, the canister might never release the resources that it used to allocate the call. Let's view an example:
+
+```
+#[update]
+async fn update_avatar(user_id: UserId, pic: ByteBuf ) {
+    let key = store_async(user_id, &pic)
+                  .await     
+                  .unwrap(); 
+    USERS.with(|users| set_avatar_key(user_id, key));
+}
+```
+
+- The method receives a byte buffer with an avatar picture.
+- The method issues a call to the storage canister. The byte buffer is captured in a Rust future allocated on the heap.
+- If the call fails, the canister panics. The system rolls back the canister state to the snapshot created right before the callback invocation. From the canister’s point of view, it still waits for the reply and keeps the future and the buffer on the heap.
+
+Note that there is no corruption, the canister is still in a valid state, but some resources, like memory, will not be released until the next upgrade.
+
+The System API was recently extended to deal with this problem (see `ic0.call_on_cleanup` in the [Internet Computer Interface Specification](../../../references/ic-interface-spec.md)). This issue is likely to be fixed in future versions of the Rust CDK.
+
+:::caution
+Do not lock shared resources across await boundaries.
+:::
+
+This issue becomes quite a problem when combined with panics: if you lock an important resource and then panic after await, the resource stays locked forever.
```

Comment:
I believe ic-cdk fixed this issue in 0.5.1.

Overall, I believe this text is based on a somewhat outdated version of https://mmapped.blog/posts/01-effective-rust-canisters.html (on a snapshot before the last October).